File: BinaryTree/Leetcode98.java
public class Leetcode98 {
    public boolean isValidBST(TreeNode root) {
        return helper(root,null,null);
    }
    
    // using this helper function we make sure that current not is in BST, and update the range for its children.
    public boolean helper(TreeNode root,Integer min,Integer max){
        if(root == null)return true;
        if(min != null && root.val <= min)return false;
        if(max != null && root.val >= max)return false;
        return helper(root.left,min,root.val) && helper(root.right,root.val,max);
    }
}

File: BinarySearchTree/Python/Leetcode98.py
class Leetcode98:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.check(root.left, None, root.val) and self.check(root.right, root.val, None) 

    def check(self, node:TreeNode, l:int, r:int) -> bool:
        if node is None:
            return True
        if l is None:
            return node.val < r and self.check(node.left, l, node.val) and self.check(node.right, node.val, r)
        if r is None:
            return l < node.val and self.check(node.left, l, node.val) and self.check(node.right, node.val, r)
        return l < node.val < r and self.check(node.left, l, node.val) and self.check(node.right, node.val, r)
